fwdctl/src/dump.rs

```diff
@@ -24,50 +28,230 @@ pub struct Options {
                 }
             }
-            Some(Err(e)) => return Err(e),
+            Err(e) => return Err(e),
         }
     }
+    output_handler.flush();
+
     Ok(())
 }
 
-fn u8_to_string(data: &[u8]) -> Cow<'_, str> {
-    String::from_utf8_lossy(data)
+const fn key_count_exceeded(max: Option<u32>, key_count: u32) -> bool {
+    matches!(max, Some(max) if key_count >= max)
+}
+
+fn u8_to_string(data: &[u8]) -> String {
+    String::from_utf8_lossy(data).to_string()
 }
 
 fn key_parser(s: &str) -> Result<Box<[u8]>, std::io::Error> {
     Ok(Box::from(s.as_bytes()))
 }
+
+// Helper function to convert key and value to a string
+fn key_value_to_string(key: &[u8], value: &[u8], hex: bool) -> (String, String) {
+    let key_str = if hex {
+        hex::encode(key)
+    } else {
+        u8_to_string(key)
+    };
+    let value_str = if hex {
+        hex::encode(value)
+    } else {
+        u8_to_string(value)
+    };
+    (key_str, value_str)
+}
+
+async fn handle_next_key(next_key: KeyFromStream) {
+    match next_key {
+        Some(Ok((key, _))) => {
+            println!(
+                "Next key is {0}, resume with \"--start-key={0}\".",
+                u8_to_string(&key)
+            );
+        }
+        Some(Err(e)) => {
+            eprintln!("Error occurred while fetching the next key: {}", e);
+        }
+        None => {
+            println!("There is no next key. Data dump completed.");
+        }
+    }
+}
+
+trait OutputHandler {
+    fn handle_record(&mut self, key: &[u8], value: &[u8]);
+    fn flush(&mut self);
+}
+
+struct CsvOutputHandler {
+    writer: csv::Writer<File>,
+    hex: bool,
+}
+
+impl OutputHandler for CsvOutputHandler {
+    fn handle_record(&mut self, key: &[u8], value: &[u8]) {
+        let (key_str, value_str) = key_value_to_string(key, value, self.hex);
+        self.writer
+            .write_record(&[key_str, value_str])
+            .expect("Failed to write CSV record");
+    }
+
+    fn flush(&mut self) {
+        self.writer.flush().expect("Failed to flush CSV writer");
+    }
+}
+
+struct JsonOutputHandler {
+    writer: BufWriter<File>,
+    hex: bool,
+    is_first: bool,
+}
+
+impl JsonOutputHandler {
+    fn new(mut writer: BufWriter<File>, hex: bool) -> JsonOutputHandler {
+        let _ = writer.write("{\n".as_bytes());
+        JsonOutputHandler {
+            writer,
+            hex,
+            is_first: true,
+        }
+    }
+}
+
+impl OutputHandler for JsonOutputHandler {
+    fn handle_record(&mut self, key: &[u8], value: &[u8]) {
+        let (key_str, value_str) = key_value_to_string(key, value, self.hex);
+        if self.is_first {
+            let _ = self
+                .writer
+                .write(format!("  \"{}\": \"{}\"", key_str, value_str).as_bytes());
+            self.is_first = false;
+        } else {
+            let _ = self
+                .writer
+                .write(format!(",\n  \"{}\": \"{}\"", key_str, value_str).as_bytes());
```

Comment:
This code is nearly identical to the other case, so I would probably separate the `is_first` case to before printing:

```rust
if self.is_first {
    self.writer.write(b",\n")?;
    self.is_first = true;
}
```